odl/space/fspace.py

```diff
@@ -154,196 +356,346 @@ class FunctionSetVector(Operator):
+            self._call_out_of_place = fcall
+        elif call_out_optional:
+            # Dual-use
+            self._call_in_place = self._call_out_of_place = fcall
+        else:
+            # In-place only
+            self._call_in_place = fcall
+            # No way to safely determine a data type for the output array,
+            # therefore not out-of-place default applicable
+            self._call_out_of_place = _out_of_place_not_impl
 
-        # Todo: allow users to specify linear
-        super().__init__(self.space.domain, self.space.range, linear=False)
+        self._vectorized = bool(vectorized)
+        if not self._vectorized and call_has_out:
+            raise ValueError('output parameters make no sense for '
+                             'non-vectorized functions.')
 
     @property
     def space(self):
-        """Return space attribute."""
+        """The space or set this function belongs to."""
         return self._space
 
-    def __eq__(self, other):
-        """Return ``self == other``.
-
-        Returns
-        -------
-        equals : `bool`
-            `True` if ``other`` is a `FunctionSetVector` with
-            ``other.space`` equal to this vector's space and
-            the call and apply implementations of ``other`` and
-            this vector are equal. `False` otherwise.
-        """
-        if other is self:
-            return True
-
-        return (isinstance(other, FunctionSetVector) and
-                self.space == other.space and
-                self._call == other._call and
-                self._apply == other._apply)
+    @property
+    def vectorized(self):
+        """Whether this function supports vectorized evaluation."""
+        return self._vectorized
+
+    def _call(self, x, out=None, **kwargs):
+        """Raw evaluation method."""
+        if out is None:
+            out = self._call_out_of_place(x, **kwargs)
+        else:
+            self._call_in_place(x, out=out, **kwargs)
+        return out
 
-    # FIXME: this is a bad hack bypassing the operator default
-    # pattern for apply and call
-    def __call__(self, *x):
-        """Vectorized and multi-argument out-of-place evaluation.
+    def __call__(self, x, out=None, **kwargs):
+        """Out-of-place evaluation.
 
         Parameters
         ----------
-        x1,...,xN : `object`
-            Input arguments for the function evaluation.
+        x : object
+            Input argument for the function evaluation. Conditions
+            on `x` depend on vectorization:
+
+            `False` : ``x`` must be a domain element
+
+            `True` : ``x`` must be a `numpy.ndarray` with shape
+            ``(d, N)``, where ``d`` is the number of dimensions of
+            the function domain
+            OR
+            `x` is a sequence of `numpy.ndarray` with length
+            `space.ndim`, and the arrays can be broadcast
+            against each other.
+
+        out : `numpy.ndarray`, optional
```

Comment:
I honestly don't know how to do it. Vectorization makes most sense for multidimensional arrays, which have no support in CUDA currently. We *could* support one-dimensional functions with CUDA vectors as output containers, I suppose.